File: leetcode/1-100/29-divide-two-integers/bench/bit_shift.py
"""Benchmark workload — bit-shift long division, LeetCode #29.

Algorithmic mirror of bench/bit_shift.kara. See ../README.md § Benchmarks for
the choice of N and the LCG input. Truncating division is re-implemented with
shifts (not Python's flooring `//`) so the sink matches the compiled mirrors.
"""

from __future__ import annotations


def divide(dividend: int, divisor: int) -> int:
    int_max = 2147483647
    int_min = -2147483648
    if dividend == int_min and divisor == -1:
        return int_max
    negative = (dividend < 0) != (divisor < 0)
    a = dividend if dividend >= 0 else -dividend
    b = divisor if divisor >= 0 else -divisor
    result = 0
    while a >= b:
        temp = b
        multiple = 1
        while a >= (temp << 1):
            temp <<= 1
            multiple <<= 1
        a -= temp
        result += multiple
    return -result if negative else result


def main() -> None:
    n = 5_000_000
    state = 1
    total = 0
    for _ in range(n):
        state = (state * 1103515245 + 12345) % 2147483648
        dividend = state - 1073741824
        state = (state * 1103515245 + 12345) % 2147483648
        divisor = (state % 2000) - 1000
        if divisor == 0:
            divisor = 1
        total += divide(dividend, divisor)
    print(total)


if __name__ == "__main__":
    main()

File: leetcode/1-100/29-divide-two-integers/bench/bit_shift.rs
//! Benchmark workload — bit-shift long division, LeetCode #29.
//!
//! Algorithmic mirror of bench/bit_shift.kara and bench/bit_shift.py. See
//! ../README.md § Benchmarks for the choice of N and the LCG input.

fn divide(dividend: i64, divisor: i64) -> i64 {
    const INT_MAX: i64 = 2147483647;
    const INT_MIN: i64 = -2147483648;
    if dividend == INT_MIN && divisor == -1 {
        return INT_MAX;
    }
    let negative = (dividend < 0) != (divisor < 0);
    let mut a = dividend.abs();
    let b = divisor.abs();
    let mut result = 0i64;
    while a >= b {
        let mut temp = b;
        let mut multiple = 1i64;
        while a >= (temp << 1) {
            temp <<= 1;
            multiple <<= 1;
        }
        a -= temp;
        result += multiple;
    }
    if negative { -result } else { result }
}

fn main() {
    const N: i64 = 5_000_000;
    let mut state: i64 = 1;
    let mut sum: i64 = 0;
    for _ in 0..N {
        state = (state * 1103515245 + 12345) % 2147483648;
        let dividend = state - 1073741824;
        state = (state * 1103515245 + 12345) % 2147483648;
        let mut divisor = (state % 2000) - 1000;
        if divisor == 0 {
            divisor = 1;
        }
        sum += divide(dividend, divisor);
    }
    println!("{}", sum);
}
